src/sched/async.c

```diff
@@ -0,0 +1,709 @@
+		ERROR_RETURN_LOG_ERRNO(int, "Cannot initialize the awaiting list mutex");
+
+	if(NULL == (_ctx.q_data = (_handle_t**)malloc(sizeof(_ctx.q_data[0]) * _ctx.q_cap)))
+		ERROR_RETURN_LOG_ERRNO(int, "Cannot allocate memory for the queue memory");
+
+	if(NULL == (_ctx.q_pool = mempool_objpool_new(sizeof(_handle_t))))
+		ERROR_LOG_GOTO(ERR, "Cannot create memory pool for async handles");
+
+	/* Initialize the awaiting list */
+	if(NULL == (_ctx.al_list = (_awaiter_t*)malloc(sizeof(_awaiter_t) * _ctx.al_cap)))
+		ERROR_LOG_GOTO(ERR, "Cannot allocate memory for the awaiting task list");
+
+	_ctx.al_unused = 0;
+
+	for(i = 0; i < _ctx.al_cap; i ++)
+	{
+		_ctx.al_list[i].valid = 0;
+		_ctx.al_list[i].next = i + 1;
+		_ctx.al_list[i].task = NULL;
+	}
+	_ctx.al_list[_ctx.al_cap - 1].next = ERROR_CODE(uint32_t);
+	_ctx.al_done = ERROR_CODE(uint32_t);
+
+	/* Then, we need to start the async processing threads */
+	if(NULL == (_ctx.thread_data = (_thread_data_t*)calloc(sizeof(_thread_data_t), _ctx.nthreads)))
+		ERROR_LOG_ERRNO_GOTO(ERR, "Cannot allocate memory for the async task thread pool");
+
+	for(i = 0; i < _ctx.nthreads; i ++)
+		if(NULL == (_ctx.thread_data[i].thread = thread_new(_async_processor_main, _ctx.thread_data + i, THREAD_TYPE_ASYNC)))
+			ERROR_LOG_GOTO(ERR, "Cannot start the new thread for the Async Task Processor");
+
+	/* Finally, we  should set the initialization flag */
+	_ctx.init = 1;
+
+	return 0;
+
+ERR:
+
+	if(_ctx.q_data != NULL) free(_ctx.q_data);
+	if(_ctx.q_pool != NULL) mempool_objpool_free(_ctx.q_pool);
+	if(_ctx.thread_data != NULL)
+	{
+		_ctx.killed = 1;
+
+		/* We need to let everyone know they are killed */
+		pthread_cond_broadcast(&_ctx.q_cond);
+
+		for(i = 0; i < _ctx.nthreads; i ++)
+			if(_ctx.thread_data[i].thread != NULL)
+				thread_free(_ctx.thread_data[i].thread, NULL);
+
+		free(_ctx.thread_data);
+	}
+	if(_ctx.al_list != NULL) free(_ctx.al_list);
+	return ERROR_CODE(int);
+}
+
+int sched_async_kill()
+{
+	if(!_ctx.init) ERROR_RETURN_LOG(int, "The async processor haven't been started yet");
+	int rc = 0;
+
+	_ctx.killed = 1;
+
+	/* Let't kill all the async processing thread at this point */
+	pthread_cond_broadcast(&_ctx.q_cond);
+	uint32_t i;
+	for(i = 0; i < _ctx.nthreads; i ++)
+		if(ERROR_CODE(int) == thread_free(_ctx.thread_data[i].thread, NULL))
+			rc = ERROR_CODE(int);
+	free(_ctx.thread_data);
+
+	/* We need to dispose the queue */
+	for(i = _ctx.q_front; i < _ctx.q_rear; i ++)
+	{
+		if(_ctx.q_data[i & (_ctx.q_cap - 1)] == NULL) continue;
+		if(_ctx.q_data[i & (_ctx.q_cap - 1)]->exec_task != NULL && 
+		   ERROR_CODE(int) == runtime_task_free(_ctx.q_data[i & (_ctx.q_cap - 1)]->exec_task))
+			rc = ERROR_CODE(int);
+		if(ERROR_CODE(int) == mempool_objpool_dealloc(_ctx.q_pool, _ctx.q_data[i & (_ctx.q_cap - 1)]))
+			rc = ERROR_CODE(int);
+	}
+
+	if(ERROR_CODE(int) == mempool_objpool_free(_ctx.q_pool))
+		rc = ERROR_CODE(int);
+
+	free(_ctx.q_data);
+
+	free(_ctx.al_list);
```

Comment:
Is it possible that there's some async handle has been dequeued and in the list. We are losing the the pending handles at this point